stock_reserve_rule/models/stock_reserve_rule.py

```diff
@@ -333,3 +398,66 @@ def _apply_strategy_full_bin(self, quants):
 
             if float_compare(need, location_quantity, rounding) != -1:
                 need = yield location, location_quantity, need, None, None
+
+    def _compare_with_tolerance(self, need, product_qty, rounding):
+        tolerance = self.tolerance_requested_value
+        limit = self.tolerance_requested_limit
+        computation = self.tolerance_requested_computation
+        if limit == "no_tolerance" or float_compare(tolerance, 0, rounding) == 0:
+            return float_compare(need, product_qty, rounding) == 0
+        elif limit == "upper_limit":
+            if computation == "percentage":
+                # need + rounding < product_qty <= need * (100 + tolerance) / 100
+                return (
+                    float_compare(need, product_qty, rounding) == -1
+                    and float_compare(
+                        product_qty, need * (100 + tolerance) / 100, rounding
+                    )
+                    <= 0
+                )
+            else:
+                # need + rounding < product_qty <= need + tolerance
+                return (
+                    float_compare(need, product_qty, rounding) == -1
+                    and float_compare(product_qty, need + tolerance, rounding) <= 0
+                )
+        elif limit == "lower_limit":
+            if computation == "percentage":
+                # need * (100 - tolerance) / 100 <= product_qty < need - rounding
+                return (
+                    float_compare(need * (100 - tolerance) / 100, product_qty, rounding)
+                    <= 0
+                    and float_compare(product_qty, need, rounding) == -1
+                )
+            # computation == "absolute"
+            else:
+                # need - tolerance <= product_qty < need - rounding
+                return (
+                    float_compare(need - tolerance, product_qty, rounding) <= 0
+                    and float_compare(product_qty, need, rounding) == -1
+                )
+
+    def _apply_strategy_full_lot(self, quants):
+        need = yield
+        # We take goods only if we empty the bin.
+        # The original ordering (fefo, fifo, ...) must be kept.
+        product = fields.first(quants).product_id
+        if product.tracking == "lot":
+            rounding = product.uom_id.rounding
+            for lot, lot_quants in quants.filtered(
+                lambda quant: quant.product_id.id == product.id
+                and quant.lot_id
+                and not quant.reserved_quantity
```

Comment:
You cannot exclude quant with reserved quantities before the group by. If the lot is on 2 locations and has a reservation only on one location, you will not end-up with the right result

Reply:
Now we lost the exclusion. See my proposed solution above